### ISSUE TITLE
Implement fault injection helpers

### ISSUE THREAD

**Issue by bitwalker** [bitwalker/ex_unit_clustered_case#2]
It's often desirable to test that some behavior in the application is consistent in the face of random faults, or that it matches the expected behavior. The nature of these faults is often application specific, though there are some faults which can be offered out of the box (i.e. random partitions).

An API is needed which supports injecting faults in useful ways, and supports arbitrary extension (some constraints may be needed, but I'm not yet sure what those are).

- [ ] First, need to figure out what types of faults apply to applications in general
- [ ] Second, determine what type of faults may be specific to a few different applications, and then find traits in common so we can work towards an API which allows both applications to implement fault injection in the ways they are needed

**Comment by beardedeagle** [comment]
A couple of suggestions as conversation starters:

- latency
- timeouts
- unreachable nodes
- flapping
- unhandled messages
- lib/app failure (supervised app failures)
- crashed procs

**Comment by beardedeagle** [comment]
I would agree with the "user defined" faults bit. As for injecting faults, I think the pieces with reasonable docs would be good enough to be honest.
